From dcm experimental implementation

adbugger opened the pull request:
This is the experimental implementation for review. This branch is not to be merged as is. Canceling builds.

adbugger committed:
ENH: Added scalar implementation of from_dcm for testing only

adbugger committed:
MAINT: Updated experimental implementation

Forgot to git add to the last commit

adbugger commented:
Forgot to git add in the last commit

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,35 @@ def as_dcm(self):
             return dcm
+
+    @classmethod
+    def from_dcm(cls, mat, scalar=True):
```
mat -> dcm

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,35 @@ def as_dcm(self):
             return dcm
+
+    @classmethod
+    def from_dcm(cls, mat, scalar=True):
```
In the final implementation we won't have `scalar` flag and the algorithm will be selected automatically, right?

adbugger replied:
No, `scalar` flag is only for now.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,35 @@ def as_dcm(self):
+        quat = np.empty((num_rotations, 4))
+        if scalar:
+            for rot_num in range(num_rotations):
+                A = mat[rot_num]
```
I think it would make sense to refactor this code into a standalone function like `_dcm_to_quaternion(dcm)` or `_dcm_to_quaternion_single(dcm)`.

adbugger replied:
Will do.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,35 @@ def as_dcm(self):
+            for rot_num in range(num_rotations):
+                A = mat[rot_num]
+                trA = A.trace()
+                choice = np.argmax(np.array([A[0, 0], A[1, 1], A[2, 2], trA]))
```
I think `np.array` is not strictly necessary here.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,35 @@ def as_dcm(self):
+                choice = np.argmax(np.array([A[0, 0], A[1, 1], A[2, 2], trA]))
+
+                if choice in [0, 1, 2]:
+                    quat[rot_num, choice] = 1 - trA + 2 * A[choice, choice]
```
I'm not 100% sure how to write this code in the "best way", but I probably would introduce 3 indices `i, j, k`, like:
```
i = choise
j = (choice + 1) % 3
k = (choice + 2) % 3
```
And then use this indices to write the formulas. The "idea" here is to avoid somewhat confusing names like `next_ind` and `next_next` and to use neutral mathematical indices. I think it will read slightly nicer. You might have another better idea.

By the way, this is a great generalization of 3 out of 4 cases.

adbugger replied:
The `i, j, k` indices look and read better. Once the implementation is fixed, I'll change the variable names.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,35 @@ def as_dcm(self):
+                    quat[rot_num, next_ind] = A[choice, next_ind] + A[next_ind, choice]
+                    quat[rot_num, next_next] = A[choice, next_next] + A[next_next, choice]
+
+                    quat[rot_num, 3] = A[next_ind, next_next] - A[next_next, next_ind]
```
I believe here you forgot that the convention in the paper is the opposite to what we adopted, so you need to transpose indices.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,35 @@ def as_dcm(self):
+                    quat[rot_num, 3] = A[next_ind, next_next] - A[next_next, next_ind]
+                else:
+                    quat[rot_num] = np.array([
+                        A[1, 2] - A[2, 1],
```
The same thing with indices transposition.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,35 @@ def as_dcm(self):
+        # For testing purposes only. Return cls(quat) in final implementation.
+        # Creating new objects in python is time consuming. We do not want to
+        # count that time in the benchmarks.
+        return quat
```
To be honest I believe there is no problem in creating the object and counting the time. It will be the same for all cases.

Another thing I'm not sure about --- maybe we should normalize `quat` here as an one liner `quat /= np.linalg.norm(quat, axis=1)[:, None]`. Because technically it will avoid copying and norm checks in `__init__`.

Considering this, maybe we should make `__init__` trivial as `self._quat = quat` and delegate `quat` preparation (as normalized 2-d array) wherever else. What do you think about that?

adbugger replied:
Yup, we could pass `normalize=True` to completely skip norm calculations and zero norm checks.

I think `__init__` is fine as is. If we remove preparation from `__init__` there is a possibility that the user may use the `__init__` method to get un-normalized quaternions. While most users will read the documentation and not use `__init__` I'd be more comfortable if we removed that possibility completely. In any case, passing `normalized=True` flag gives us the same functionality and avoids norm calculations  and checks, so I think it's fine as is. Does my reasoning sound sensible? Maybe I'm on the wrong track here.

nmayorov replied:
I think you are right, no reason to modify `__init__`. My reasoning was to remove `__init__` as a an initialization method for a user completely, sort of "cleaner" design. But I suggest to keep it. Maybe we will reconsider something later.

nmayorov reviewed (commented):
Reviewed the preliminarily version.

nmayorov commented:
@adbugger what about vectorized approach, how is it going?

adbugger commented:
I coded the vectorized approach and I'm ironing out the bugs right now. It took me a while to figure out how to vectorize it and I'm not sure if I have the best possible vectorized code yet. I'll push that, along with the benchmarks in by tonight's scheduled update. The scalar experiments are already in the wiki, by the way.

nmayorov commented:
Great, waiting for the update.

In your experiments, how did you generated matrices? We definitely need to have them random (but close to orthogonal I guess) so each of 4 cases appears uniformly.

adbugger commented:
For the experiments, why do the generated matrices need to be orthogonal?

I am generating random matrices. Here we are just checking a general method for converting a dcm (well, any 3x3 matrix in general) into a quaternion. Orthogonal matrices produce orthogonal quaternions, definitely, but for benchmarking purposes, aren't random matrices enough? From the paper, the computations performed are independent of the orthogonality of the input matrix.

adbugger committed:
ENH: Experimental implementation of vectorized approach

nmayorov commented:
I think you can generate random matrices (although how exactly?). But can you be sure that all 4 cases will appear with approximately equal frequencies? I don't mind any method, however please check that all cases occur with equal probability.

adbugger commented:
I was generating random matrices with the `np.random.random` function.
But you're right, I haven't  ensured that all the cases occur with equal probability. How would I ensure that?

nmayorov commented:
I have two ideas:

1. Perhaps that everything is good with the current method. Easy to check.
2. Generate a random rotation uniformly --- 4 random normal numbers and then `Rotation.from_quaterion` to them. Then get dcm by `as_dcm`. You can add small noise to it afterwards. Actually, I think you should do that. Because this is only a guess about equal probabilities of 4 cases --- in reality we need random rotations (interesting to see whether 4 cases will be equiprobable).

adbugger commented:
Your second point is intriguing. Generating 4 random numbers will definitely get us random rotations, but will they give us the 4 cases with equal probability?

If the 4 cases were symmetric upto a cyclic permutation (like the first three are), I'd be confident in saying a straightforward yes.
However since the fourth value, the trace, depends on the other three (randomly generated) values, I'm not sure that will generate the 4 cases with equal probability.
In any case, benchmarking with random rotations seems the best idea.

nmayorov commented:
> Your second point is intriguing. Generating 4 random numbers will definitely get us random rotations, but will they give us the 4 cases with equal probability?

Let me explain a bit more. I think that in fact we need to test on random rotations generated uniformly over all possible rotations (over SO3 group or whatever). My reasoning is that on average we probably will get a set of random rotations and that we should evaluate the performance on that. 

Also notice that you need to generated 4 numbers as normal (Gaussian) random variables, after normalizing you will get a point uniformly over 4D sphere --- and this is what we need.

adbugger committed:
MAINT: Fixed array indexing

adbugger commented:
@nmayorov the wiki has been updated with the latest benchmarks.

What I can see is that there is a huge performance penalty for vectorized operations on a single rotation; almost 4 times. However for number of rotations > 7, vectorized approach is better.

Right now, the vectorized approach computes all the possible quaternions and chooses amongst them based on the cases. In effect it does 4 times the computation of the scalar case and still beats it for more than 7 rotations. If we make the vector implementation better, there might be performance improvements.

What do you suggest we do from here?

nmayorov commented:
@adbugger I will give my opinion/review before you start working today.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,56 @@ def as_dcm(self):
+                        A[1, 0] - A[0, 1],
+                        1 + trA])
+        else:
+            decision_matrix = np.zeros((num_rotations, 4))
```
I think it's not the optimal way to implement that. I suggest the following approach.

1. You computed array of choices --- this is good.
2. Create the return array `quat = np.empty((n_rotations, 4))`.
3. Consider cases. 
```
ind = np.nonzero(choices != 3)[0]
i = choices[ind]
j = (i + 1) % 3
k = (j + 1) % 3
# Fill the quat[ind, 0], quat[ind, 1] in the same way as in scalar approach, use mat[ind, ...].

ind = np.nonzero(choices == 3)[0]
# Do the similar
```

nmayorov commented:
@adbugger so consider the approach for vectorized computations I suggested. Then rerun the benchmarks. I suspect that in reality a scalar approach won't be necessary, but we'll see. Then start implementing a clean version and tests.

larsoner commented:
If the vectorized version is faster for anything bigger than ~10 (I think you said 7 in your blog post?) let's just always use that. It's simpler than triaging, and maintainability is something we must keep in mind.

There is a quote: "Premature optimization is the root of all evil" so I'd vote to go with the vectorized approach here and optimize the scalar approach later if we really need to. It's reasonable for Python folks familiar with optimization to expect creating many (thousands to see a big slowdown?) `Rotation` objects to be slower than creating one `Rotation` object that contains many transforms itself. We can make a note in the docstring that putting multiple rotations in a single object is faster than creating multiple objects.

adbugger commented:
The wiki has been updated with the latest timings. The vector implementation now catches up at 5 rotations, and there is a much better performance benefit of using the vector implementation now.

I'll get onto the cleaner version, documenting and testing.

adbugger committed:
MAINT: Faster vector implementation.

The benchmarks on the wiki has been updated with the latest timings
and the vector implementation performs much better than the scalar
one. Moving on to cleaner implementation, documentation, and testing.

nmayorov commented:
@larsoner It is possible that you want to update a single rotation iteratively and compute dcm from it on each step (attitude kinematics problem). For this scenario I think it is beneficial to optimize operations for a single rotation.

I agree that this is borderline, but since it is already implemented I'm sort of +0.5 to keep both approaches. 

@adbugger, @larsoner what would be your score for this in [-1, +1]. I think it makes sense to make consensus decision based on this kind of voting )

adbugger commented:
It's already implemented so there's no effort cost involved in keeping both. I'm just not sure how important of a use case it is. Hence I'll say +0.25

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -129,3 +129,56 @@ def as_dcm(self):
+    def from_dcm(cls, mat, scalar=True):
+        # For now, mat must be a 3D matrix
+        num_rotations = mat.shape[0]
+        if scalar:
```
If we are keeping both methods, how would we organize the code?

1. Two separate self-contained functions?
2. Keep as it is written now?

I think 2 would be better. My original suggestion to factor out some logic into separate function --- I don't like it.

adbugger replied:
Even I prefer keeping it as it is written. If we need the functionality later on in another function, then we can decide to separate out some common logic (although at this point I'm not sure if we'll require that)

larsoner commented:
> It's already implemented so there's no effort cost involved in keeping both.

There is added maintenance cost. Any test you implement for one method needs to be implemented for the other. They need to be checked for equivalence. And any future changes / updates to one needs to think about potential effects on the other.

Unless we are really convinced that we need it by use cases that will be common for people and impact their workflow speed meaningfully, -1 on including the scalar approach due to maintainability concerns.

adbugger committed:
MAINT : Cleaned up code and decided on vector implementation.

nmayorov commented:
> There is added maintenance cost. Any test you implement for one method needs to be implemented for the other. They need to be checked for equivalence. And any future changes / updates to one needs to think about potential effects on the other.

> Unless we are really convinced that we need it by use cases that will be common for people and impact their workflow speed meaningfully, -1 on including the scalar approach due to maintainability concerns.

These are good points. I think the best strategy is to release a version with vectorized approach only. After all we can't get a very good performance in iterative procedures in pure Python anyway.

adbugger committed:
TST: Added tests for from_dcm function

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -131,54 +132,58 @@ def as_dcm(self):
+
+        This classmethod return a `Rotation` object from the input direction
+        cosine matrices. If the input matrix is not orthogonal, this method
+        creates the best approximation of the rotation.
```
Just a note for future: I wouldn't called it "best" and the reference to the paper is the most appropriate here.

adbugger replied:
The paper currently lives in my google drive folder, which I don't think is appropriate to include here as a link. Can you give me a link to your original source of the paper? I'll put it here.

nmayorov replied:
The journal information at the top of the first page. 

As how to write references, see for example here https://github.com/scipy/scipy/blob/dfb0cc2/scipy/interpolate/_cubic.py#L352-L770

adbugger commented:
Build fails with `python3.4.2` and `numpy==1.8.2`. I'm using `np.matmul` in one of my tests. Looks like it wasn't introduced in `numpy==1.8.2`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -131,54 +132,58 @@ def as_dcm(self):
+        Parameters
+        ----------
+        dcm : array_like, shape (N, 3, 3) or (3, 3)
+            Each `dcm[i]` corresponds to a `(3, 3)` direction cosine matrix.
```
I would probably write: "A single matrix or a stack of matrices, where `dcm[i]` is  an i-th matrix." Or we can leave polishing docstrings for later.

adbugger replied:
I'll make this change now itself.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -131,54 +132,58 @@ def as_dcm(self):
+        quat[ind, 2] = dcm[ind, 1, 0] - dcm[ind, 0, 1]
+        quat[ind, 3] = 1 + decision_matrix[ind, -1]
+
+        if is_single:
```
Perhaps do the normalization here and pass with `normalized=True`?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -131,54 +132,58 @@ def as_dcm(self):
+
+        num_rotations = dcm.shape[0]
+
+        decision_matrix = np.zeros((num_rotations, 4))
```
This is totally minor, but technically you can use `empty` here. I think it makes sense logically as you fill each element explicitly afterwards.

adbugger replied:
I'm filling out `decision_matrix[:, :3]` explicitly, and the the last column containing the trace is calculated by summing each row `decision_matrix[:, -1] = decision_matrix.sum(axis=1)`. For this to work, we need the last column to have zeros initially.

nmayorov replied:
Right, it's sort of tricky then. Perhaps rewrite as `decision_matrix[:, :3].sum(axis=1)`?

adbugger replied:
Sure.

nmayorov reviewed (commented):
Besides some minor suggestions it looks good to me.

As for the better documentation --- I don't mind postponing it for later when we will have a full picture in front of us.

adbugger committed:
TST: Updated test to remove use of np.matmul

The matmul function was not introduced in numpy v1.8.2 and causes
travis-ci builds to fail. Instead of checking product of matrix and
transpose with identity, now we check equality of matrix inverse and
its transpose.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -190,11 +190,8 @@ def test_dcm_calculation_pipeline():
-    assert_array_almost_equal(result, eye3d)
+    # numpy.__matmul__ was not present in numpy v1.8.2. Avoid use
+    assert_array_almost_equal(
+            np.linalg.inv(ortho_dcm),
```
To be honest, I'm not very fond of using matrix inversion here. Instead of matmul, try to use https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.einsum.html

It should be something like `...ij,...jk->...ik` for the signature.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -140,3 +141,57 @@ def test_as_dcm_from_generic_input():
+
+
+def test_dcm_calculation_pipeline():
+    dcm = np.array([special_ortho_group.rvs(3) for _ in range(10)])
```
It's better to fix the random state. So it can be rewritten as `rvs(3, size=10, random_state=0)`.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -140,3 +141,57 @@ def test_as_dcm_from_generic_input():
+
+
+def test_from_dcm_ortho_output():
+    dcm = np.random.random((100, 3, 3))
```
Also fix the random seed, with `np.random.seed(0)`.

adbugger committed:
MAINT: normalized=True and np.empty in from_dcm

The `from_dcm` function now passes normalized quaternions to `__init__`
and uses np.empty to initialize `decision_matrix`.

adbugger committed:
TST: Fixed random state and random seed in tests

adbugger committed:
DOC: Updated documentation of from_dcm and added reference

adbugger committed:
TST: Used np.einsum instead of matmul

Travis-CI has been setup to test on a build with numpy==1.8.2, which
did not have np.matmul. To ensure test cases pass, Einstein summation
convention has been used to multiply matrices.

nmayorov committed:
MAINT: Normalize quat in a single place in Rotation.from_dcm

nmayorov commented:
Made a small modification and merging.
